backend/routes.js

```diff
@@ -247,3 +248,97 @@ route.GET('/app/{path*}', {}, {
 route.GET('/', {}, function (req, h) {
   return h.redirect('/app/')
 })
+
+route.POST('/zkpp/{contract}', {
+  validate: {
+    payload: Joi.alternatives([
+      {
+        b: Joi.string().required()
+      },
+      {
+        r: Joi.string().required(),
+        s: Joi.string().required(),
+        sig: Joi.string().required(),
+        Eh: Joi.string().required()
+      }
+    ])
+  }
+}, async function (req, h) {
+  if (req.payload['b']) {
+    const result = await registrationKey(req.params['contract'], req.payload['b'])
+
+    if (!result) {
+      return Boom.internal('internal error')
+    }
+
+    return result
+  } else {
+    const result = await register(req.params['contract'], req.payload['r'], req.payload['s'], req.payload['sig'], req.payload['Eh'])
+
+    if (!result) {
+      return Boom.internal('internal error')
+    }
+
+    return result
+  }
+})
+
+route.GET('/zkpp/{contract}/auth_hash', {}, async function (req, h) {
+  if (!req.query['b']) {
+    return Boom.badRequest('b query param required')
+  }
+
+  const challenge = await getChallenge(req.params['contract'], req.query['b'])
+
+  if (!challenge) {
+    return Boom.internal('internal error')
+  }
+
+  return challenge
+})
+
+route.GET('/zkpp/{contract}/contract_hash', {}, async function (req, h) {
+  if (!req.query['r']) {
+    return Boom.badRequest('r query param required')
+  }
+
+  if (!req.query['s']) {
+    return Boom.badRequest('s query param required')
+  }
+
+  if (!req.query['sig']) {
+    return Boom.badRequest('sig query param required')
+  }
+
+  if (!req.query['hc']) {
+    return Boom.badRequest('hc query param required')
+  }
+
+  const salt = await getContractSalt(req.params['contract'], req.query['r'], req.query['s'], req.query['sig'], req.query['hc'])
+
+  if (!salt) {
+    return Boom.internal('internal error')
```

Comment:
Hm, this error happens when the salt cannot be fetched (because it didn't exist or because the protocol wasn't executed correctly). I'm a bit torn about what to do in this situation, as having a 'debug' mode can be useful, but in general we shouldn't fill up the logs when someone maliciously executes the protocol, especially if that can result in timing information. Should I add some log entry like 'unable to get contract salt'?

Reply:
>  we shouldn't fill up the logs when someone maliciously executes the protocol, especially if that can result in timing information. Should I add some log entry like 'unable to get contract salt'?

If an error occurs we should always log it, and especially if the error could occur because of malicious activity. This is essential for things like `fail2ban`. We should also log it in a consistent way so that fail2ban rules can be written to extract the IP.